File: hello.py
#!/usr/bin/env python3

import os
import sys


def read_body():
    try:
        length = int(os.environ.get("CONTENT_LENGTH", "0") or "0")
    except ValueError:
        length = 0
    if length <= 0:
        return ""
    data = sys.stdin.buffer.read(length)
    return data.decode("utf-8", errors="replace")


def escape(text):
    return (text.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;"))


def main():
    method = os.environ.get("REQUEST_METHOD", "")
    query = os.environ.get("QUERY_STRING", "")
    remote_addr = os.environ.get("REMOTE_ADDR", "?")
    server_name = os.environ.get("SERVER_NAME", "?")
    server_port = os.environ.get("SERVER_PORT", "?")
    script_name = os.environ.get("SCRIPT_NAME", "?")
    path_info = os.environ.get("PATH_INFO", "?")
    user_agent = os.environ.get("HTTP_USER_AGENT", "(none)")
    host_header = os.environ.get("HTTP_HOST", "(none)")

    body = read_body() if method == "POST" else ""

    http_vars = sorted(k for k in os.environ if k.startswith("HTTP_"))

    print("Content-Type: text/html")
    print()
    print("<!DOCTYPE html><html><head><title>webserv CGI</title>"
          "<link rel=\"stylesheet\" href=\"/style.css\"></head><body>")
    print("<h1>Hello from CGI</h1>")
    print("<section><h2>Request</h2><ul>")
    print("<li>REQUEST_METHOD: <code>%s</code></li>" % escape(method))
    print("<li>SCRIPT_NAME: <code>%s</code></li>" % escape(script_name))
    print("<li>PATH_INFO: <code>%s</code></li>" % escape(path_info))
    print("<li>QUERY_STRING: <code>%s</code></li>" % escape(query))
    print("<li>REMOTE_ADDR: <code>%s</code></li>" % escape(remote_addr))
    print("<li>SERVER_NAME:SERVER_PORT: <code>%s:%s</code></li>" % (escape(server_name), escape(server_port)))
    print("<li>Host header: <code>%s</code></li>" % escape(host_header))
    print("<li>User-Agent: <code>%s</code></li>" % escape(user_agent))
    print("</ul></section>")

    if method == "POST":
        print("<section><h2>Body received (%d bytes)</h2><pre>%s</pre></section>"
              % (len(body), escape(body)))

    print("<section><h2>All HTTP_* env vars seen by the CGI</h2><ul>")
    for key in http_vars:
        print("<li><code>%s</code> = <code>%s</code></li>" % (escape(key), escape(os.environ[key])))
    print("</ul></section>")
    print("<p><a href=\"/\">&larr; back</a></p>")
    print("</body></html>")


if __name__ == "__main__":
    main()

File: hello.js
#!/usr/bin/env node
// hello.py'nin JavaScript (node) karsiligi -- webserv'in birden fazla CGI
// dilini ayni anda destekledigini gostermek icin. Ayni CGI/1.1 sozlesmesi:
// baglam env'den, POST body stdin'den, cikti "header'lar + bos satir + body".
"use strict";

function escapeHtml(s) {
    return String(s).replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;");
}

function readBody(cb) {
    var length = parseInt(process.env.CONTENT_LENGTH || "0", 10) || 0;
    if (length <= 0 || process.env.REQUEST_METHOD !== "POST")
        return cb("");
    var chunks = [];
    var got = 0;
    process.stdin.on("data", function (chunk) {
        chunks.push(chunk);
        got += chunk.length;
        if (got >= length)
            process.stdin.pause();
    });
    process.stdin.on("end", function () {});
    process.stdin.on("pause", flush);
    process.stdin.on("close", flush);
    var done = false;
    function flush() {
        if (done) return;
        done = true;
        cb(Buffer.concat(chunks).toString("utf8"));
    }
}

readBody(function (body) {
    var env = process.env;
    var out = [];
    out.push("Content-Type: text/html");
    out.push("");
    out.push("<!DOCTYPE html><html><head><title>webserv CGI (js)</title>" +
             '<link rel="stylesheet" href="/style.css"></head><body>');
    out.push("<h1>Hello from CGI (JavaScript)</h1>");
    out.push("<section><h2>Request</h2><ul>");
    out.push("<li>REQUEST_METHOD: <code>" + escapeHtml(env.REQUEST_METHOD || "") + "</code></li>");
    out.push("<li>SCRIPT_NAME: <code>" + escapeHtml(env.SCRIPT_NAME || "?") + "</code></li>");
    out.push("<li>QUERY_STRING: <code>" + escapeHtml(env.QUERY_STRING || "") + "</code></li>");
    out.push("<li>SERVER_NAME:SERVER_PORT: <code>" +
             escapeHtml(env.SERVER_NAME || "?") + ":" + escapeHtml(env.SERVER_PORT || "?") + "</code></li>");
    out.push("<li>User-Agent: <code>" + escapeHtml(env.HTTP_USER_AGENT || "(none)") + "</code></li>");
    out.push("</ul></section>");
    if (env.REQUEST_METHOD === "POST")
        out.push("<section><h2>Body received (" + body.length + " bytes)</h2><pre>" +
                 escapeHtml(body) + "</pre></section>");
    out.push('<p><a href="/">&larr; back</a></p>');
    out.push("</body></html>");
    process.stdout.write(out.join("\n") + "\n");
});
